{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is a Python Script to manipulte the output files of orthofinder for plotting or further analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd \n",
    "import os\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '/home/jake/Downloads/HSP_Project/Results/Comparative_Genomics_Statistics/Duplications_per_Orthogroup.tsv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m/home/jake/Downloads/HSP_Project/Exploring-the-Evoltuoin-of-Heatshock-Proteins-Across-the-Diversity-of-Fish/scripts/Data_Manipulation.ipynb Cell 3\u001b[0m in \u001b[0;36m1\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/jake/Downloads/HSP_Project/Exploring-the-Evoltuoin-of-Heatshock-Proteins-Across-the-Diversity-of-Fish/scripts/Data_Manipulation.ipynb#W2sZmlsZQ%3D%3D?line=8'>9</a>\u001b[0m Single_Copy_Orthologue_Sequences \u001b[39m=\u001b[39m \u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39m{\u001b[39;00mpath\u001b[39m}\u001b[39;00m\u001b[39msingle_Copy_Orthologue_Sequences/\u001b[39m\u001b[39m'\u001b[39m\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/jake/Downloads/HSP_Project/Exploring-the-Evoltuoin-of-Heatshock-Proteins-Across-the-Diversity-of-Fish/scripts/Data_Manipulation.ipynb#W2sZmlsZQ%3D%3D?line=11'>12</a>\u001b[0m species_Taxonomy \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mread_csv(\u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39m{\u001b[39;00mpath\u001b[39m}\u001b[39;00m\u001b[39mSpecies_Taxonomy.csv\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[0;32m---> <a href='vscode-notebook-cell:/home/jake/Downloads/HSP_Project/Exploring-the-Evoltuoin-of-Heatshock-Proteins-Across-the-Diversity-of-Fish/scripts/Data_Manipulation.ipynb#W2sZmlsZQ%3D%3D?line=12'>13</a>\u001b[0m Duplications_per_Orthogroup \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39;49mread_csv(\u001b[39mf\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39m{\u001b[39;49;00mComparative_Genomics_Statistics\u001b[39m}\u001b[39;49;00m\u001b[39mDuplications_per_Orthogroup.tsv\u001b[39;49m\u001b[39m'\u001b[39;49m, sep \u001b[39m=\u001b[39;49m \u001b[39m'\u001b[39;49m\u001b[39m\\t\u001b[39;49;00m\u001b[39m'\u001b[39;49m, index_col\u001b[39m=\u001b[39;49m\u001b[39m0\u001b[39;49m)\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/jake/Downloads/HSP_Project/Exploring-the-Evoltuoin-of-Heatshock-Proteins-Across-the-Diversity-of-Fish/scripts/Data_Manipulation.ipynb#W2sZmlsZQ%3D%3D?line=13'>14</a>\u001b[0m Duplications \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mread_csv(\u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39m{\u001b[39;00mGene_Duplication_Events\u001b[39m}\u001b[39;00m\u001b[39mDuplications.tsv\u001b[39m\u001b[39m'\u001b[39m, sep \u001b[39m=\u001b[39m \u001b[39m'\u001b[39m\u001b[39m\\t\u001b[39;00m\u001b[39m'\u001b[39m, index_col\u001b[39m=\u001b[39m\u001b[39m0\u001b[39m)\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/jake/Downloads/HSP_Project/Exploring-the-Evoltuoin-of-Heatshock-Proteins-Across-the-Diversity-of-Fish/scripts/Data_Manipulation.ipynb#W2sZmlsZQ%3D%3D?line=14'>15</a>\u001b[0m OrthologuesStats_Totals \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mread_csv(\u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39m{\u001b[39;00mComparative_Genomics_Statistics\u001b[39m}\u001b[39;00m\u001b[39mOrthologuesStats_Totals.tsv\u001b[39m\u001b[39m'\u001b[39m, sep \u001b[39m=\u001b[39m \u001b[39m'\u001b[39m\u001b[39m\\t\u001b[39;00m\u001b[39m'\u001b[39m, index_col\u001b[39m=\u001b[39m\u001b[39m0\u001b[39m)\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/util/_decorators.py:311\u001b[0m, in \u001b[0;36mdeprecate_nonkeyword_arguments.<locals>.decorate.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    305\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mlen\u001b[39m(args) \u001b[39m>\u001b[39m num_allow_args:\n\u001b[1;32m    306\u001b[0m     warnings\u001b[39m.\u001b[39mwarn(\n\u001b[1;32m    307\u001b[0m         msg\u001b[39m.\u001b[39mformat(arguments\u001b[39m=\u001b[39marguments),\n\u001b[1;32m    308\u001b[0m         \u001b[39mFutureWarning\u001b[39;00m,\n\u001b[1;32m    309\u001b[0m         stacklevel\u001b[39m=\u001b[39mstacklevel,\n\u001b[1;32m    310\u001b[0m     )\n\u001b[0;32m--> 311\u001b[0m \u001b[39mreturn\u001b[39;00m func(\u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/io/parsers/readers.py:678\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, error_bad_lines, warn_bad_lines, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options)\u001b[0m\n\u001b[1;32m    663\u001b[0m kwds_defaults \u001b[39m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m    664\u001b[0m     dialect,\n\u001b[1;32m    665\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    674\u001b[0m     defaults\u001b[39m=\u001b[39m{\u001b[39m\"\u001b[39m\u001b[39mdelimiter\u001b[39m\u001b[39m\"\u001b[39m: \u001b[39m\"\u001b[39m\u001b[39m,\u001b[39m\u001b[39m\"\u001b[39m},\n\u001b[1;32m    675\u001b[0m )\n\u001b[1;32m    676\u001b[0m kwds\u001b[39m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m--> 678\u001b[0m \u001b[39mreturn\u001b[39;00m _read(filepath_or_buffer, kwds)\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/io/parsers/readers.py:575\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    572\u001b[0m _validate_names(kwds\u001b[39m.\u001b[39mget(\u001b[39m\"\u001b[39m\u001b[39mnames\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39mNone\u001b[39;00m))\n\u001b[1;32m    574\u001b[0m \u001b[39m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 575\u001b[0m parser \u001b[39m=\u001b[39m TextFileReader(filepath_or_buffer, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwds)\n\u001b[1;32m    577\u001b[0m \u001b[39mif\u001b[39;00m chunksize \u001b[39mor\u001b[39;00m iterator:\n\u001b[1;32m    578\u001b[0m     \u001b[39mreturn\u001b[39;00m parser\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/io/parsers/readers.py:932\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m    929\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39moptions[\u001b[39m\"\u001b[39m\u001b[39mhas_index_names\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m=\u001b[39m kwds[\u001b[39m\"\u001b[39m\u001b[39mhas_index_names\u001b[39m\u001b[39m\"\u001b[39m]\n\u001b[1;32m    931\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mhandles: IOHandles \u001b[39m|\u001b[39m \u001b[39mNone\u001b[39;00m \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m\n\u001b[0;32m--> 932\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_engine \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_make_engine(f, \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mengine)\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/io/parsers/readers.py:1216\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1212\u001b[0m     mode \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mrb\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m   1213\u001b[0m \u001b[39m# error: No overload variant of \"get_handle\" matches argument types\u001b[39;00m\n\u001b[1;32m   1214\u001b[0m \u001b[39m# \"Union[str, PathLike[str], ReadCsvBuffer[bytes], ReadCsvBuffer[str]]\"\u001b[39;00m\n\u001b[1;32m   1215\u001b[0m \u001b[39m# , \"str\", \"bool\", \"Any\", \"Any\", \"Any\", \"Any\", \"Any\"\u001b[39;00m\n\u001b[0;32m-> 1216\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mhandles \u001b[39m=\u001b[39m get_handle(  \u001b[39m# type: ignore[call-overload]\u001b[39;49;00m\n\u001b[1;32m   1217\u001b[0m     f,\n\u001b[1;32m   1218\u001b[0m     mode,\n\u001b[1;32m   1219\u001b[0m     encoding\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49moptions\u001b[39m.\u001b[39;49mget(\u001b[39m\"\u001b[39;49m\u001b[39mencoding\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39mNone\u001b[39;49;00m),\n\u001b[1;32m   1220\u001b[0m     compression\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49moptions\u001b[39m.\u001b[39;49mget(\u001b[39m\"\u001b[39;49m\u001b[39mcompression\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39mNone\u001b[39;49;00m),\n\u001b[1;32m   1221\u001b[0m     memory_map\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49moptions\u001b[39m.\u001b[39;49mget(\u001b[39m\"\u001b[39;49m\u001b[39mmemory_map\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39mFalse\u001b[39;49;00m),\n\u001b[1;32m   1222\u001b[0m     is_text\u001b[39m=\u001b[39;49mis_text,\n\u001b[1;32m   1223\u001b[0m     errors\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49moptions\u001b[39m.\u001b[39;49mget(\u001b[39m\"\u001b[39;49m\u001b[39mencoding_errors\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39m\"\u001b[39;49m\u001b[39mstrict\u001b[39;49m\u001b[39m\"\u001b[39;49m),\n\u001b[1;32m   1224\u001b[0m     storage_options\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49moptions\u001b[39m.\u001b[39;49mget(\u001b[39m\"\u001b[39;49m\u001b[39mstorage_options\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39mNone\u001b[39;49;00m),\n\u001b[1;32m   1225\u001b[0m )\n\u001b[1;32m   1226\u001b[0m \u001b[39massert\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mhandles \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m\n\u001b[1;32m   1227\u001b[0m f \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mhandles\u001b[39m.\u001b[39mhandle\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/io/common.py:786\u001b[0m, in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    781\u001b[0m \u001b[39melif\u001b[39;00m \u001b[39misinstance\u001b[39m(handle, \u001b[39mstr\u001b[39m):\n\u001b[1;32m    782\u001b[0m     \u001b[39m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[1;32m    783\u001b[0m     \u001b[39m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[1;32m    784\u001b[0m     \u001b[39mif\u001b[39;00m ioargs\u001b[39m.\u001b[39mencoding \u001b[39mand\u001b[39;00m \u001b[39m\"\u001b[39m\u001b[39mb\u001b[39m\u001b[39m\"\u001b[39m \u001b[39mnot\u001b[39;00m \u001b[39min\u001b[39;00m ioargs\u001b[39m.\u001b[39mmode:\n\u001b[1;32m    785\u001b[0m         \u001b[39m# Encoding\u001b[39;00m\n\u001b[0;32m--> 786\u001b[0m         handle \u001b[39m=\u001b[39m \u001b[39mopen\u001b[39;49m(\n\u001b[1;32m    787\u001b[0m             handle,\n\u001b[1;32m    788\u001b[0m             ioargs\u001b[39m.\u001b[39;49mmode,\n\u001b[1;32m    789\u001b[0m             encoding\u001b[39m=\u001b[39;49mioargs\u001b[39m.\u001b[39;49mencoding,\n\u001b[1;32m    790\u001b[0m             errors\u001b[39m=\u001b[39;49merrors,\n\u001b[1;32m    791\u001b[0m             newline\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39m\"\u001b[39;49m,\n\u001b[1;32m    792\u001b[0m         )\n\u001b[1;32m    793\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    794\u001b[0m         \u001b[39m# Binary mode\u001b[39;00m\n\u001b[1;32m    795\u001b[0m         handle \u001b[39m=\u001b[39m \u001b[39mopen\u001b[39m(handle, ioargs\u001b[39m.\u001b[39mmode)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/home/jake/Downloads/HSP_Project/Results/Comparative_Genomics_Statistics/Duplications_per_Orthogroup.tsv'"
     ]
    }
   ],
   "source": [
    "# Read in the all the Datafiles from the Orthofinder output\n",
    "path = '/home/jake/Downloads/HSP_Project/Results/'\n",
    "Comparative_Genomics_Statistics = f'{path}Comparative_Genomics_Statistics/'\n",
    "Gene_Duplication_Events = f'{path}Gene_Duplication_Events/'\n",
    "Orthogroups = f'{path}Orthogroups/'\n",
    "Orthologues = f'{path}Orthologues/'\n",
    "Phylogenetically_Misplaced_Genes = f'{path}Phylogenetically_Misplaced_Genes/'\n",
    "Putative_Xenologs = f'{path}Putative_Xenologs/'\n",
    "Single_Copy_Orthologue_Sequences = f'{path}single_Copy_Orthologue_Sequences/'\n",
    "\n",
    "\n",
    "species_Taxonomy = pd.read_csv(f'{path}Species_Taxonomy.csv')\n",
    "Duplications_per_Orthogroup = pd.read_csv(f'{Comparative_Genomics_Statistics}Duplications_per_Orthogroup.tsv', sep = '\\t', index_col=0)\n",
    "Duplications = pd.read_csv(f'{Gene_Duplication_Events}Duplications.tsv', sep = '\\t', index_col=0)\n",
    "OrthologuesStats_Totals = pd.read_csv(f'{Comparative_Genomics_Statistics}OrthologuesStats_Totals.tsv', sep = '\\t', index_col=0)\n",
    "StatisticsPerSpecies = pd.read_csv(f'{Comparative_Genomics_Statistics}Statistics_PerSpecies.tsv', sep = '\\t', index_col=0)\n",
    "Statistics_Overall = pd.read_csv(f'{Comparative_Genomics_Statistics}Statistics_Overall.tsv', sep = '\\t', nrows = 16, index_col=0)\n",
    "Orthogroups = pd.read_csv(f'{Orthogroups}Orthogroups.tsv', sep = '\\t', index_col=0)\n",
    "#Orthogroups_GeneCount= pd.read_csv(f'{Orthogroups}Orthogroups.GeneCount.tsv', sep = '\\t', index_col=0)\\\n",
    "\n",
    "file = '/home/jake/Downloads/HSP_Project/headerz/headerz.txt'\n",
    "#filtered_headers = headers[headers[0].str.contains(\"Heat|HSP|Hsp|heat|shps|hsp|HEAT|shock|\", na=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Acanthochromis_polyacanthus</th>\n",
       "      <th>Amphilophus_citrinellus</th>\n",
       "      <th>Amphiprion_ocellaris</th>\n",
       "      <th>Amphiprion_percula</th>\n",
       "      <th>Anabas_testudineus</th>\n",
       "      <th>Astatotilapia_calliptera</th>\n",
       "      <th>Astyanax_mexicanus</th>\n",
       "      <th>Betta_splendens</th>\n",
       "      <th>Callorhinchus_milii</th>\n",
       "      <th>Carassius_auratus</th>\n",
       "      <th>...</th>\n",
       "      <th>Sinocyclocheilus_anshuiensis</th>\n",
       "      <th>Sinocyclocheilus_grahami</th>\n",
       "      <th>Sinocyclocheilus_rhinocerous</th>\n",
       "      <th>Sparus_aurata</th>\n",
       "      <th>Sphaeramia_orbicularis</th>\n",
       "      <th>Stegastes_partitus</th>\n",
       "      <th>Takifugu_rubripes</th>\n",
       "      <th>Tetraodon_nigroviridis</th>\n",
       "      <th>Xiphophorus_couchianus</th>\n",
       "      <th>Xiphophorus_maculatus</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Orthogroup</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OG0000000</th>\n",
       "      <td>ENSAPOG00000001294.1, ENSAPOG00000005221.1, EN...</td>\n",
       "      <td>ENSACIG00000000133.1, ENSACIG00000001635.1, EN...</td>\n",
       "      <td>ENSAOCG00000001306.1, ENSAOCG00000001404.1, EN...</td>\n",
       "      <td>ENSAPEG00000000175.1, ENSAPEG00000000226.1, EN...</td>\n",
       "      <td>ENSATEG00000002820.2, ENSATEG00000003160.2, EN...</td>\n",
       "      <td>ENSACLG00000000836.1, ENSACLG00000001397.1, EN...</td>\n",
       "      <td>ENSAMXG00005000517.1, ENSAMXG00005000543.1, EN...</td>\n",
       "      <td>ENSBSLG00000000122.1, ENSBSLG00000000302.1, EN...</td>\n",
       "      <td>ENSCMIG00000010370.1</td>\n",
       "      <td>ENSCARG00000000097.1, ENSCARG00000000098.1, EN...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSSANG00000000644.1, ENSSANG00000002744.1, EN...</td>\n",
       "      <td>ENSSGRG00000000845.1, ENSSGRG00000001446.1, EN...</td>\n",
       "      <td>ENSSRHG00000000584.1, ENSSRHG00000001351.1, EN...</td>\n",
       "      <td>ENSSAUG00010000123.1, ENSSAUG00010000129.1, EN...</td>\n",
       "      <td>ENSSORG00005000054.1, ENSSORG00005000200.1, EN...</td>\n",
       "      <td>ENSSPAG00000000196.1, ENSSPAG00000000228.1, EN...</td>\n",
       "      <td>ENSTRUG00000000100.3, ENSTRUG00000001325.3, EN...</td>\n",
       "      <td>ENSTNIG00000000305.1, ENSTNIG00000000406.1, EN...</td>\n",
       "      <td>ENSXCOG00000000161.1, ENSXCOG00000000181.1, EN...</td>\n",
       "      <td>ENSXMAG00000001829.2, ENSXMAG00000005175.2, EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0000001</th>\n",
       "      <td>ENSAPOG00000000480.1, ENSAPOG00000001531.1, EN...</td>\n",
       "      <td>ENSACIG00000000441.1, ENSACIG00000000978.1, EN...</td>\n",
       "      <td>ENSAOCG00000000175.1, ENSAOCG00000001612.1, EN...</td>\n",
       "      <td>ENSAPEG00000001723.1, ENSAPEG00000001927.1, EN...</td>\n",
       "      <td>ENSATEG00000001224.2, ENSATEG00000001882.2, EN...</td>\n",
       "      <td>ENSACLG00000000306.1, ENSACLG00000000324.1, EN...</td>\n",
       "      <td>ENSAMXG00005000282.1, ENSAMXG00005000531.1, EN...</td>\n",
       "      <td>ENSBSLG00000001116.1, ENSBSLG00000001734.1, EN...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSCARG00000000099.1, ENSCARG00000000176.1, EN...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSSANG00000000463.1, ENSSANG00000016961.1, EN...</td>\n",
       "      <td>ENSSGRG00000004788.1, ENSSGRG00000006812.1, EN...</td>\n",
       "      <td>ENSSRHG00000005690.1, ENSSRHG00000005880.1, EN...</td>\n",
       "      <td>ENSSAUG00010001237.1, ENSSAUG00010003535.1, EN...</td>\n",
       "      <td>ENSSORG00005002371.1, ENSSORG00005002998.1, EN...</td>\n",
       "      <td>ENSSPAG00000000440.1, ENSSPAG00000001161.1, EN...</td>\n",
       "      <td>ENSTRUG00000026037.1, ENSTRUG00000026133.1, EN...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSXCOG00000010994.1, ENSXCOG00000014876.1, EN...</td>\n",
       "      <td>ENSXMAG00000021132.1, ENSXMAG00000021223.1, EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0000002</th>\n",
       "      <td>ENSAPOG00000000050.1, ENSAPOG00000002022.1, EN...</td>\n",
       "      <td>ENSACIG00000000322.1, ENSACIG00000000339.1, EN...</td>\n",
       "      <td>ENSAOCG00000000720.1, ENSAOCG00000001143.1, EN...</td>\n",
       "      <td>ENSAPEG00000000925.1, ENSAPEG00000002107.1, EN...</td>\n",
       "      <td>ENSATEG00000000550.2, ENSATEG00000000837.2, EN...</td>\n",
       "      <td>ENSACLG00000000165.1, ENSACLG00000000224.1, EN...</td>\n",
       "      <td>ENSAMXG00005000185.1, ENSAMXG00005000414.1, EN...</td>\n",
       "      <td>ENSBSLG00000000744.1, ENSBSLG00000000748.1, EN...</td>\n",
       "      <td>ENSCMIG00000000854.1, ENSCMIG00000000925.1, EN...</td>\n",
       "      <td>ENSCARG00000001624.1, ENSCARG00000001841.1, EN...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSSANG00000000310.1, ENSSANG00000000350.1, EN...</td>\n",
       "      <td>ENSSGRG00000001005.1, ENSSGRG00000001501.1, EN...</td>\n",
       "      <td>ENSSRHG00000000183.1, ENSSRHG00000005213.1, EN...</td>\n",
       "      <td>ENSSAUG00010002043.1, ENSSAUG00010003182.1, EN...</td>\n",
       "      <td>ENSSORG00005001433.1, ENSSORG00005001434.1, EN...</td>\n",
       "      <td>ENSSPAG00000000781.1, ENSSPAG00000001130.1, EN...</td>\n",
       "      <td>ENSTRUG00000001904.3, ENSTRUG00000003968.3, EN...</td>\n",
       "      <td>ENSTNIG00000001084.1, ENSTNIG00000008570.1, EN...</td>\n",
       "      <td>ENSXCOG00000000187.1, ENSXCOG00000000323.1, EN...</td>\n",
       "      <td>ENSXMAG00000001098.2, ENSXMAG00000002798.2, EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0000003</th>\n",
       "      <td>ENSAPOG00000003533.1, ENSAPOG00000010330.1, EN...</td>\n",
       "      <td>ENSACIG00000000038.1, ENSACIG00000008842.1, EN...</td>\n",
       "      <td>ENSAOCG00000004576.1, ENSAOCG00000011105.1, EN...</td>\n",
       "      <td>ENSAPEG00000000917.1, ENSAPEG00000001164.1, EN...</td>\n",
       "      <td>ENSATEG00000002707.2, ENSATEG00000002739.2, EN...</td>\n",
       "      <td>ENSACLG00000000182.1, ENSACLG00000000230.1, EN...</td>\n",
       "      <td>ENSAMXG00005009960.1, ENSAMXG00005011923.1</td>\n",
       "      <td>ENSBSLG00000000758.1, ENSBSLG00000000761.1, EN...</td>\n",
       "      <td>ENSCMIG00000000204.1, ENSCMIG00000000243.1, EN...</td>\n",
       "      <td>ENSCARG00000000106.1, ENSCARG00000000580.1, EN...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSSANG00000046842.1</td>\n",
       "      <td>ENSSGRG00000002355.1, ENSSGRG00000002662.1, EN...</td>\n",
       "      <td>ENSSRHG00000000933.1, ENSSRHG00000002281.1, EN...</td>\n",
       "      <td>ENSSAUG00010000151.1, ENSSAUG00010000690.1, EN...</td>\n",
       "      <td>ENSSORG00005000217.1, ENSSORG00005000700.1, EN...</td>\n",
       "      <td>ENSSPAG00000005749.1, ENSSPAG00000005778.1, EN...</td>\n",
       "      <td>ENSTRUG00000024195.2, ENSTRUG00000024564.2, EN...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSXCOG00000003154.1, ENSXCOG00000015327.1</td>\n",
       "      <td>ENSXMAG00000019451.2, ENSXMAG00000020915.1, EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0000004</th>\n",
       "      <td>ENSAPOG00000005592.1, ENSAPOG00000006503.1, EN...</td>\n",
       "      <td>ENSACIG00000000255.1, ENSACIG00000000460.1, EN...</td>\n",
       "      <td>ENSAOCG00000001381.1, ENSAOCG00000001597.1, EN...</td>\n",
       "      <td>ENSAPEG00000006320.1, ENSAPEG00000006377.1, EN...</td>\n",
       "      <td>ENSATEG00000007083.2, ENSATEG00000007152.2, EN...</td>\n",
       "      <td>ENSACLG00000008503.1, ENSACLG00000008522.1, EN...</td>\n",
       "      <td>ENSAMXG00005001568.1, ENSAMXG00005001594.1, EN...</td>\n",
       "      <td>ENSBSLG00000006294.1, ENSBSLG00000013599.1, EN...</td>\n",
       "      <td>ENSCMIG00000000068.1, ENSCMIG00000000069.1, EN...</td>\n",
       "      <td>ENSCARG00000005352.1, ENSCARG00000005365.1, EN...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSSANG00000000618.1, ENSSANG00000000634.1, EN...</td>\n",
       "      <td>ENSSGRG00000001520.1, ENSSGRG00000002043.1, EN...</td>\n",
       "      <td>ENSSRHG00000000525.1, ENSSRHG00000000691.1, EN...</td>\n",
       "      <td>ENSSAUG00010000174.1, ENSSAUG00010000184.1, EN...</td>\n",
       "      <td>ENSSORG00005004204.1, ENSSORG00005004209.1, EN...</td>\n",
       "      <td>ENSSPAG00000001110.1, ENSSPAG00000001546.1, EN...</td>\n",
       "      <td>ENSTRUG00000002636.3, ENSTRUG00000002672.3, EN...</td>\n",
       "      <td>ENSTNIG00000000043.1, ENSTNIG00000000109.1, EN...</td>\n",
       "      <td>ENSXCOG00000000348.1, ENSXCOG00000000749.1, EN...</td>\n",
       "      <td>ENSXMAG00000019913.2, ENSXMAG00000021075.1, EN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 85 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  Acanthochromis_polyacanthus  \\\n",
       "Orthogroup                                                      \n",
       "OG0000000   ENSAPOG00000001294.1, ENSAPOG00000005221.1, EN...   \n",
       "OG0000001   ENSAPOG00000000480.1, ENSAPOG00000001531.1, EN...   \n",
       "OG0000002   ENSAPOG00000000050.1, ENSAPOG00000002022.1, EN...   \n",
       "OG0000003   ENSAPOG00000003533.1, ENSAPOG00000010330.1, EN...   \n",
       "OG0000004   ENSAPOG00000005592.1, ENSAPOG00000006503.1, EN...   \n",
       "\n",
       "                                      Amphilophus_citrinellus  \\\n",
       "Orthogroup                                                      \n",
       "OG0000000   ENSACIG00000000133.1, ENSACIG00000001635.1, EN...   \n",
       "OG0000001   ENSACIG00000000441.1, ENSACIG00000000978.1, EN...   \n",
       "OG0000002   ENSACIG00000000322.1, ENSACIG00000000339.1, EN...   \n",
       "OG0000003   ENSACIG00000000038.1, ENSACIG00000008842.1, EN...   \n",
       "OG0000004   ENSACIG00000000255.1, ENSACIG00000000460.1, EN...   \n",
       "\n",
       "                                         Amphiprion_ocellaris  \\\n",
       "Orthogroup                                                      \n",
       "OG0000000   ENSAOCG00000001306.1, ENSAOCG00000001404.1, EN...   \n",
       "OG0000001   ENSAOCG00000000175.1, ENSAOCG00000001612.1, EN...   \n",
       "OG0000002   ENSAOCG00000000720.1, ENSAOCG00000001143.1, EN...   \n",
       "OG0000003   ENSAOCG00000004576.1, ENSAOCG00000011105.1, EN...   \n",
       "OG0000004   ENSAOCG00000001381.1, ENSAOCG00000001597.1, EN...   \n",
       "\n",
       "                                           Amphiprion_percula  \\\n",
       "Orthogroup                                                      \n",
       "OG0000000   ENSAPEG00000000175.1, ENSAPEG00000000226.1, EN...   \n",
       "OG0000001   ENSAPEG00000001723.1, ENSAPEG00000001927.1, EN...   \n",
       "OG0000002   ENSAPEG00000000925.1, ENSAPEG00000002107.1, EN...   \n",
       "OG0000003   ENSAPEG00000000917.1, ENSAPEG00000001164.1, EN...   \n",
       "OG0000004   ENSAPEG00000006320.1, ENSAPEG00000006377.1, EN...   \n",
       "\n",
       "                                           Anabas_testudineus  \\\n",
       "Orthogroup                                                      \n",
       "OG0000000   ENSATEG00000002820.2, ENSATEG00000003160.2, EN...   \n",
       "OG0000001   ENSATEG00000001224.2, ENSATEG00000001882.2, EN...   \n",
       "OG0000002   ENSATEG00000000550.2, ENSATEG00000000837.2, EN...   \n",
       "OG0000003   ENSATEG00000002707.2, ENSATEG00000002739.2, EN...   \n",
       "OG0000004   ENSATEG00000007083.2, ENSATEG00000007152.2, EN...   \n",
       "\n",
       "                                     Astatotilapia_calliptera  \\\n",
       "Orthogroup                                                      \n",
       "OG0000000   ENSACLG00000000836.1, ENSACLG00000001397.1, EN...   \n",
       "OG0000001   ENSACLG00000000306.1, ENSACLG00000000324.1, EN...   \n",
       "OG0000002   ENSACLG00000000165.1, ENSACLG00000000224.1, EN...   \n",
       "OG0000003   ENSACLG00000000182.1, ENSACLG00000000230.1, EN...   \n",
       "OG0000004   ENSACLG00000008503.1, ENSACLG00000008522.1, EN...   \n",
       "\n",
       "                                           Astyanax_mexicanus  \\\n",
       "Orthogroup                                                      \n",
       "OG0000000   ENSAMXG00005000517.1, ENSAMXG00005000543.1, EN...   \n",
       "OG0000001   ENSAMXG00005000282.1, ENSAMXG00005000531.1, EN...   \n",
       "OG0000002   ENSAMXG00005000185.1, ENSAMXG00005000414.1, EN...   \n",
       "OG0000003          ENSAMXG00005009960.1, ENSAMXG00005011923.1   \n",
       "OG0000004   ENSAMXG00005001568.1, ENSAMXG00005001594.1, EN...   \n",
       "\n",
       "                                              Betta_splendens  \\\n",
       "Orthogroup                                                      \n",
       "OG0000000   ENSBSLG00000000122.1, ENSBSLG00000000302.1, EN...   \n",
       "OG0000001   ENSBSLG00000001116.1, ENSBSLG00000001734.1, EN...   \n",
       "OG0000002   ENSBSLG00000000744.1, ENSBSLG00000000748.1, EN...   \n",
       "OG0000003   ENSBSLG00000000758.1, ENSBSLG00000000761.1, EN...   \n",
       "OG0000004   ENSBSLG00000006294.1, ENSBSLG00000013599.1, EN...   \n",
       "\n",
       "                                          Callorhinchus_milii  \\\n",
       "Orthogroup                                                      \n",
       "OG0000000                                ENSCMIG00000010370.1   \n",
       "OG0000001                                                 NaN   \n",
       "OG0000002   ENSCMIG00000000854.1, ENSCMIG00000000925.1, EN...   \n",
       "OG0000003   ENSCMIG00000000204.1, ENSCMIG00000000243.1, EN...   \n",
       "OG0000004   ENSCMIG00000000068.1, ENSCMIG00000000069.1, EN...   \n",
       "\n",
       "                                            Carassius_auratus  ...  \\\n",
       "Orthogroup                                                     ...   \n",
       "OG0000000   ENSCARG00000000097.1, ENSCARG00000000098.1, EN...  ...   \n",
       "OG0000001   ENSCARG00000000099.1, ENSCARG00000000176.1, EN...  ...   \n",
       "OG0000002   ENSCARG00000001624.1, ENSCARG00000001841.1, EN...  ...   \n",
       "OG0000003   ENSCARG00000000106.1, ENSCARG00000000580.1, EN...  ...   \n",
       "OG0000004   ENSCARG00000005352.1, ENSCARG00000005365.1, EN...  ...   \n",
       "\n",
       "                                 Sinocyclocheilus_anshuiensis  \\\n",
       "Orthogroup                                                      \n",
       "OG0000000   ENSSANG00000000644.1, ENSSANG00000002744.1, EN...   \n",
       "OG0000001   ENSSANG00000000463.1, ENSSANG00000016961.1, EN...   \n",
       "OG0000002   ENSSANG00000000310.1, ENSSANG00000000350.1, EN...   \n",
       "OG0000003                                ENSSANG00000046842.1   \n",
       "OG0000004   ENSSANG00000000618.1, ENSSANG00000000634.1, EN...   \n",
       "\n",
       "                                     Sinocyclocheilus_grahami  \\\n",
       "Orthogroup                                                      \n",
       "OG0000000   ENSSGRG00000000845.1, ENSSGRG00000001446.1, EN...   \n",
       "OG0000001   ENSSGRG00000004788.1, ENSSGRG00000006812.1, EN...   \n",
       "OG0000002   ENSSGRG00000001005.1, ENSSGRG00000001501.1, EN...   \n",
       "OG0000003   ENSSGRG00000002355.1, ENSSGRG00000002662.1, EN...   \n",
       "OG0000004   ENSSGRG00000001520.1, ENSSGRG00000002043.1, EN...   \n",
       "\n",
       "                                 Sinocyclocheilus_rhinocerous  \\\n",
       "Orthogroup                                                      \n",
       "OG0000000   ENSSRHG00000000584.1, ENSSRHG00000001351.1, EN...   \n",
       "OG0000001   ENSSRHG00000005690.1, ENSSRHG00000005880.1, EN...   \n",
       "OG0000002   ENSSRHG00000000183.1, ENSSRHG00000005213.1, EN...   \n",
       "OG0000003   ENSSRHG00000000933.1, ENSSRHG00000002281.1, EN...   \n",
       "OG0000004   ENSSRHG00000000525.1, ENSSRHG00000000691.1, EN...   \n",
       "\n",
       "                                                Sparus_aurata  \\\n",
       "Orthogroup                                                      \n",
       "OG0000000   ENSSAUG00010000123.1, ENSSAUG00010000129.1, EN...   \n",
       "OG0000001   ENSSAUG00010001237.1, ENSSAUG00010003535.1, EN...   \n",
       "OG0000002   ENSSAUG00010002043.1, ENSSAUG00010003182.1, EN...   \n",
       "OG0000003   ENSSAUG00010000151.1, ENSSAUG00010000690.1, EN...   \n",
       "OG0000004   ENSSAUG00010000174.1, ENSSAUG00010000184.1, EN...   \n",
       "\n",
       "                                       Sphaeramia_orbicularis  \\\n",
       "Orthogroup                                                      \n",
       "OG0000000   ENSSORG00005000054.1, ENSSORG00005000200.1, EN...   \n",
       "OG0000001   ENSSORG00005002371.1, ENSSORG00005002998.1, EN...   \n",
       "OG0000002   ENSSORG00005001433.1, ENSSORG00005001434.1, EN...   \n",
       "OG0000003   ENSSORG00005000217.1, ENSSORG00005000700.1, EN...   \n",
       "OG0000004   ENSSORG00005004204.1, ENSSORG00005004209.1, EN...   \n",
       "\n",
       "                                           Stegastes_partitus  \\\n",
       "Orthogroup                                                      \n",
       "OG0000000   ENSSPAG00000000196.1, ENSSPAG00000000228.1, EN...   \n",
       "OG0000001   ENSSPAG00000000440.1, ENSSPAG00000001161.1, EN...   \n",
       "OG0000002   ENSSPAG00000000781.1, ENSSPAG00000001130.1, EN...   \n",
       "OG0000003   ENSSPAG00000005749.1, ENSSPAG00000005778.1, EN...   \n",
       "OG0000004   ENSSPAG00000001110.1, ENSSPAG00000001546.1, EN...   \n",
       "\n",
       "                                            Takifugu_rubripes  \\\n",
       "Orthogroup                                                      \n",
       "OG0000000   ENSTRUG00000000100.3, ENSTRUG00000001325.3, EN...   \n",
       "OG0000001   ENSTRUG00000026037.1, ENSTRUG00000026133.1, EN...   \n",
       "OG0000002   ENSTRUG00000001904.3, ENSTRUG00000003968.3, EN...   \n",
       "OG0000003   ENSTRUG00000024195.2, ENSTRUG00000024564.2, EN...   \n",
       "OG0000004   ENSTRUG00000002636.3, ENSTRUG00000002672.3, EN...   \n",
       "\n",
       "                                       Tetraodon_nigroviridis  \\\n",
       "Orthogroup                                                      \n",
       "OG0000000   ENSTNIG00000000305.1, ENSTNIG00000000406.1, EN...   \n",
       "OG0000001                                                 NaN   \n",
       "OG0000002   ENSTNIG00000001084.1, ENSTNIG00000008570.1, EN...   \n",
       "OG0000003                                                 NaN   \n",
       "OG0000004   ENSTNIG00000000043.1, ENSTNIG00000000109.1, EN...   \n",
       "\n",
       "                                       Xiphophorus_couchianus  \\\n",
       "Orthogroup                                                      \n",
       "OG0000000   ENSXCOG00000000161.1, ENSXCOG00000000181.1, EN...   \n",
       "OG0000001   ENSXCOG00000010994.1, ENSXCOG00000014876.1, EN...   \n",
       "OG0000002   ENSXCOG00000000187.1, ENSXCOG00000000323.1, EN...   \n",
       "OG0000003          ENSXCOG00000003154.1, ENSXCOG00000015327.1   \n",
       "OG0000004   ENSXCOG00000000348.1, ENSXCOG00000000749.1, EN...   \n",
       "\n",
       "                                        Xiphophorus_maculatus  \n",
       "Orthogroup                                                     \n",
       "OG0000000   ENSXMAG00000001829.2, ENSXMAG00000005175.2, EN...  \n",
       "OG0000001   ENSXMAG00000021132.1, ENSXMAG00000021223.1, EN...  \n",
       "OG0000002   ENSXMAG00000001098.2, ENSXMAG00000002798.2, EN...  \n",
       "OG0000003   ENSXMAG00000019451.2, ENSXMAG00000020915.1, EN...  \n",
       "OG0000004   ENSXMAG00000019913.2, ENSXMAG00000021075.1, EN...  \n",
       "\n",
       "[5 rows x 85 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(Orthogroups.head(5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "headers = pd.read_csv(file, sep='\\t', header=None, names=['header'])\n",
    "# extract gene up until the first space\n",
    "headers['gene'] = headers['header'].str.extract(r'gene:(.*?)\\s').fillna('')\n",
    "\n",
    "# extract description\n",
    "headers['description'] = headers['header'].str.extract(r'description:(.*)\\[').fillna('')\n",
    "#Drop the rows that have nothing in the description column\n",
    "headers = headers[headers['description'] != '']\n",
    "headers = headers[['gene', 'description']]\n",
    "headers_df = headers[['gene', 'description']].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Change the dtype of the columns to string\n",
    "headers_df = headers[['gene', 'description']].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# I want to filter the Orthogroups df to only include the rows where 95% of the columns are not nan\n",
    "\n",
    "#First I need to get the number of columns in the df\n",
    "num_cols = len(Orthogroups.columns)\n",
    "\n",
    "#Then I need to get the number of rows in the df\n",
    "num_rows = len(Orthogroups.index)\n",
    "\n",
    "#Then I need to get the number of rows that have 95% of the columns filled\n",
    "num_rows_95 = num_rows * 0.95\n",
    "\n",
    "#Then I need to get the number of columns that have 95% of the rows filled\n",
    "num_cols_95 = num_cols * 0.95\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_fasta_to_df(file, symbol):\n",
    "    fasta_list = []\n",
    "    current_header = None\n",
    "    current_sequence = \"\"\n",
    "    with open(file, 'r') as f:\n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            if line[0] == symbol:\n",
    "                if current_header:\n",
    "                    fasta_list.append([current_header, current_sequence])\n",
    "                current_header = line[1:]\n",
    "                current_sequence = \"\"\n",
    "            else:\n",
    "                current_sequence += line.strip()\n",
    "    return pd.DataFrame(fasta_list, columns=['header', 'sequence'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>header</th>\n",
       "      <th>sequence</th>\n",
       "      <th>prefix</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSAPOG00000020882.1</td>\n",
       "      <td>MSKGPAVGIDLGTTYSCVGIFQHGKVEIIANDQGNRTTPSYVAFTD...</td>\n",
       "      <td>ENSAPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSAPOG00000020874.1</td>\n",
       "      <td>MFSCSVGVFKNGRVEIIANDQGNRITPSYVAFTSEGERLIGDAAKN...</td>\n",
       "      <td>ENSAPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSAPOG00000020752.1</td>\n",
       "      <td>MFSCSVGVFKNGRVEIIANDQGNRITPSYVAFTSEGERLIGDAAKN...</td>\n",
       "      <td>ENSAPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSAPOG00000016429.1</td>\n",
       "      <td>SSVHILSEGNASSTLTSKMKLLWVILVLVIGTVFADDYDKKENVGT...</td>\n",
       "      <td>ENSAPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSAPOG00000016402.1</td>\n",
       "      <td>SPAIGVACRCTNCVSVFKNGRVEIIANDQGNRITPSYVAFTSEGER...</td>\n",
       "      <td>ENSAPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>618</th>\n",
       "      <td>ENSXMAG00000020191.2</td>\n",
       "      <td>MSSAKGISIGIDLGTTYSCVGVFQHGKVEIIANDQGNRTTPSYVAF...</td>\n",
       "      <td>ENSXMA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>619</th>\n",
       "      <td>ENSXMAG00000001728.2</td>\n",
       "      <td>MSKGPAVGIDLGTTYSCVGVFQHGKVEIIANDQGNRTTPSYVAFTD...</td>\n",
       "      <td>ENSXMA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>620</th>\n",
       "      <td>ENSXMAG00000010572.2</td>\n",
       "      <td>MSKGPAVGIDLGTTYSCVGIFQHGKVEIIANDQGNRTTPSYVAFTD...</td>\n",
       "      <td>ENSXMA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>621</th>\n",
       "      <td>ENSXMAG00000021898.1</td>\n",
       "      <td>MCAAKNVAIGIDLGTTYSCVGVFQHGKVEIIANDQGNRTTPSYVAF...</td>\n",
       "      <td>ENSXMA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>622</th>\n",
       "      <td>ENSXMAG00000010876.2</td>\n",
       "      <td>MKSKNVAIGIDLGTTYSCVGVFQHGKVEIIANDQGNRTTPSYVAFT...</td>\n",
       "      <td>ENSXMA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>623 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   header                                           sequence  \\\n",
       "0    ENSAPOG00000020882.1  MSKGPAVGIDLGTTYSCVGIFQHGKVEIIANDQGNRTTPSYVAFTD...   \n",
       "1    ENSAPOG00000020874.1  MFSCSVGVFKNGRVEIIANDQGNRITPSYVAFTSEGERLIGDAAKN...   \n",
       "2    ENSAPOG00000020752.1  MFSCSVGVFKNGRVEIIANDQGNRITPSYVAFTSEGERLIGDAAKN...   \n",
       "3    ENSAPOG00000016429.1  SSVHILSEGNASSTLTSKMKLLWVILVLVIGTVFADDYDKKENVGT...   \n",
       "4    ENSAPOG00000016402.1  SPAIGVACRCTNCVSVFKNGRVEIIANDQGNRITPSYVAFTSEGER...   \n",
       "..                    ...                                                ...   \n",
       "618  ENSXMAG00000020191.2  MSSAKGISIGIDLGTTYSCVGVFQHGKVEIIANDQGNRTTPSYVAF...   \n",
       "619  ENSXMAG00000001728.2  MSKGPAVGIDLGTTYSCVGVFQHGKVEIIANDQGNRTTPSYVAFTD...   \n",
       "620  ENSXMAG00000010572.2  MSKGPAVGIDLGTTYSCVGIFQHGKVEIIANDQGNRTTPSYVAFTD...   \n",
       "621  ENSXMAG00000021898.1  MCAAKNVAIGIDLGTTYSCVGVFQHGKVEIIANDQGNRTTPSYVAF...   \n",
       "622  ENSXMAG00000010876.2  MKSKNVAIGIDLGTTYSCVGVFQHGKVEIIANDQGNRTTPSYVAFT...   \n",
       "\n",
       "     prefix  \n",
       "0    ENSAPO  \n",
       "1    ENSAPO  \n",
       "2    ENSAPO  \n",
       "3    ENSAPO  \n",
       "4    ENSAPO  \n",
       "..      ...  \n",
       "618  ENSXMA  \n",
       "619  ENSXMA  \n",
       "620  ENSXMA  \n",
       "621  ENSXMA  \n",
       "622  ENSXMA  \n",
       "\n",
       "[623 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fa = '/home/jake/Downloads/HSP_Project/Results/Orthogroup_Sequences/OG0000151.fa'\n",
    "fa_df = parse_fasta_to_df(fa, '>')\n",
    "#Create a new collumn that splits the header \n",
    "for col in fa_df.header:\n",
    "    fa_df['prefix'] = fa_df['header'].str.split('G0').str[0]\n",
    "display(fa_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'temp' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [15], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[39m#Join the two dataframes on the prefix column\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m dfx \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mmerge(fa_df, temp, on\u001b[39m=\u001b[39m\u001b[39m'\u001b[39m\u001b[39mprefix\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[1;32m      5\u001b[0m dfx[\u001b[39m'\u001b[39m\u001b[39mnew_head\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m=\u001b[39m dfx[\u001b[39m'\u001b[39m\u001b[39mSpecies\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m+\u001b[39m \u001b[39m\"\u001b[39m\u001b[39m:\u001b[39m\u001b[39m\"\u001b[39m \u001b[39m+\u001b[39m dfx[\u001b[39m'\u001b[39m\u001b[39mheader\u001b[39m\u001b[39m'\u001b[39m]\n\u001b[1;32m      7\u001b[0m \u001b[39m# Make a new data frame that has only the new_head and sequence columns\u001b[39;00m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'temp' is not defined"
     ]
    }
   ],
   "source": [
    "#Join the two dataframes on the prefix column\n",
    "dfx = pd.merge(fa_df, temp, on='prefix')\n",
    "\n",
    "\n",
    "dfx['new_head'] = dfx['Species'] + \":\" + dfx['header']\n",
    "\n",
    "# Make a new data frame that has only the new_head and sequence columns\n",
    "dfx = dfx[['new_head', 'sequence']]\n",
    "\n",
    "\n",
    "with open('/home/jake/Downloads/OG0000151_modified.fasta', 'w') as f:\n",
    "    for i, row in dfx.iterrows():\n",
    "        header = row['new_head']\n",
    "        sequence = row['sequence']\n",
    "        f.write(f\">{header}\\n{sequence}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#filtered_df = modded_df[modded_df['description'].str.contains('heat|hsp|HSP|shps|HEAT|shock', case=False)]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "e7370f93d1d0cde622a1f8e1c04877d8463912d04d973331ad4851f04de6915a"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
